Benchmarks/rt threads cpp

chengronglai opened the pull request:
Enhance and Simplify Real-time Threads Benchmarking

Overview
This PR significantly improves the real-time threads benchmarking tool by consolidating it to a single C++ implementation while enhancing functionality and usability. The changes provide a streamlined experience with automatic plot generation and comprehensive real-time scheduling analysis.

Key Changes
- Move from Python implementation to C++ implementation
- Improved Configuration
  - Updated load-intensity to separate bg-load-intensity/bg-workload-size and bm-load-intensity/bm-workload-size
  - Added argument for specifying dummy load parameters
- Update Output
  - Focus on Frame Period Standard Deviation in output
  - Reformatted output logs for better readability
- Updated documentation

<!-- greptile_comment -->

<h2>Greptile Overview</h2>

Updated On: 2025-10-20 14:23:35 UTC

### Greptile Summary

This PR replaces the Python-based real-time thread benchmarking tool with a C++ implementation to provide lower overhead, more precise timing measurements, and integrated plotting capabilities. The implementation creates a Holoscan application with benchmark and dummy load operators that measure frame period consistency under different Linux scheduling policies (normal, SCHED_DEADLINE, SCHED_FIFO, SCHED_RR). The configuration has been refactored to separate background load parameters (`bg-load-intensity`, `bg-workload-size`) from benchmark load parameters (`bm-load-intensity`, `bm-workload-size`), providing finer-grained control over test conditions. The tool outputs JSON-formatted statistics focusing on frame period standard deviation, then automatically invokes a Python plotting script to generate visualization graphs. This change aligns with Holoscan's C++ core and provides direct access to real-time scheduling primitives without Python GIL overhead.

<details>
<summary>Changed Files</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| benchmarks/realtime_threads_benchmarking/realtime_thread_benchmark.cpp | 2/5 | New C++ benchmark implementation with critical issues: compiler-optimizable dummy workload, potential thread safety issues in vector access, hardcoded file path using `__FILE__` that breaks on relocation, and division-by-zero risk in statistics calculation |
| benchmarks/realtime_threads_benchmarking/CMakeLists.txt | 4/5 | Complete rewrite to build C++ executable instead of Python script; includes hardcoded Holoscan SDK paths and copies plotting script without runtime dependency validation |
| benchmarks/realtime_threads_benchmarking/README.md | 4/5 | Documentation updated to reflect C++ implementation benefits, new configuration parameters, simplified metrics output, and reformatted examples |
| benchmarks/realtime_threads_benchmarking/plot_results.py | 4/5 | New Python plotting script for visualizing C++ benchmark JSON output; generates timing and histogram plots with proper error handling |
| benchmarks/realtime_threads_benchmarking/realtime_thread_benchmark.py | 5/5 | Complete deletion of original Python implementation as part of consolidation to C++ |
| benchmarks/realtime_threads_benchmarking/metadata.json | 4/5 | Updated language from Python to cpp and changed run command to compiled binary; version remains at 1.0 despite significant implementation change |

</details>

### Confidence score: 2/5

- This PR contains critical implementation issues that will prevent the benchmark from functioning correctly and may cause runtime failures
- Score reflects serious problems: the dummy workload can be optimized away by the compiler (rendering background load ineffective), unsynchronized vector access creates thread safety issues, hardcoded `__FILE__` path resolution will fail when the binary is relocated, division-by-zero risk in statistics calculation, and previous reviewer concerns about command-line parsing and thread count detection remain unaddressed
- Pay close attention to benchmarks/realtime_threads_benchmarking/realtime_thread_benchmark.cpp (lines 87-98 for workload optimization, lines 189-193 for thread safety, lines 599-600 for division-by-zero, lines 627-628 for path resolution); CMakeLists.txt also needs review for hardcoded paths and missing Python dependency validation

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant main
    participant NonRT_App as RealtimeThreadBenchmarkApp<br/>(Non-RT)
    participant BenchmarkOp_NonRT as BenchmarkOp<br/>(Baseline)
    participant DummyLoadOps_NonRT as DummyLoadOps<br/>(Background Load)
    participant RT_App as RealtimeThreadBenchmarkApp<br/>(Real-time)
    participant BenchmarkOp_RT as BenchmarkOp<br/>(Real-time)
    participant DummyLoadOps_RT as DummyLoadOps<br/>(Background Load)
    participant PlotScript as plot_results.py

    User->>main: "Run benchmark with parameters"
    main->>main: "Parse command line arguments"
    main->>main: "Print benchmark configuration"
    
    Note over main,DummyLoadOps_NonRT: Baseline Run (Non-RT Scheduling)
    main->>NonRT_App: "Create and configure (no RT scheduling)"
    main->>NonRT_App: "run()"
    NonRT_App->>NonRT_App: "compose()"
    NonRT_App->>BenchmarkOp_NonRT: "Create BenchmarkOp"
    NonRT_App->>DummyLoadOps_NonRT: "Create DummyLoadOps (CPU contention)"
    
    loop For duration_seconds
        NonRT_App->>BenchmarkOp_NonRT: "compute() - periodic"
        BenchmarkOp_NonRT->>BenchmarkOp_NonRT: "Record start time"
        BenchmarkOp_NonRT->>BenchmarkOp_NonRT: "Calculate frame period"
        BenchmarkOp_NonRT->>BenchmarkOp_NonRT: "run_dummy_cpu_workload()"
        BenchmarkOp_NonRT->>BenchmarkOp_NonRT: "Record execution time"
        NonRT_App->>DummyLoadOps_NonRT: "compute() - competing load"
        DummyLoadOps_NonRT->>DummyLoadOps_NonRT: "run_dummy_cpu_workload()"
    end
    
    main->>NonRT_App: "get_period_benchmark_stats()"
    NonRT_App->>BenchmarkOp_NonRT: "get_period_benchmark_stats()"
    BenchmarkOp_NonRT-->>main: "Return baseline statistics"
    main->>NonRT_App: "get_execution_time_benchmark_stats()"
    NonRT_App->>BenchmarkOp_NonRT: "get_execution_time_benchmark_stats()"
    BenchmarkOp_NonRT-->>main: "Return baseline exec times"
    
    Note over main,DummyLoadOps_RT: Real-time Run (RT Scheduling)
    main->>RT_App: "Create and configure (with RT scheduling)"
    main->>RT_App: "run()"
    RT_App->>RT_App: "compose()"
    RT_App->>BenchmarkOp_RT: "Create BenchmarkOp on RT thread pool"
    RT_App->>DummyLoadOps_RT: "Create DummyLoadOps (CPU contention)"
    
    loop For duration_seconds
        RT_App->>BenchmarkOp_RT: "compute() - periodic (RT priority)"
        BenchmarkOp_RT->>BenchmarkOp_RT: "Record start time"
        BenchmarkOp_RT->>BenchmarkOp_RT: "Calculate frame period"
        BenchmarkOp_RT->>BenchmarkOp_RT: "run_dummy_cpu_workload()"
        BenchmarkOp_RT->>BenchmarkOp_RT: "Record execution time"
        RT_App->>DummyLoadOps_RT: "compute() - competing load (normal priority)"
        DummyLoadOps_RT->>DummyLoadOps_RT: "run_dummy_cpu_workload()"
    end
    
    main->>RT_App: "get_period_benchmark_stats()"
    RT_App->>BenchmarkOp_RT: "get_period_benchmark_stats()"
    BenchmarkOp_RT-->>main: "Return RT statistics"
    main->>RT_App: "get_execution_time_benchmark_stats()"
    RT_App->>BenchmarkOp_RT: "get_execution_time_benchmark_stats()"
    BenchmarkOp_RT-->>main: "Return RT exec times"
    
    Note over main,PlotScript: Results and Visualization
    main->>main: "Print benchmark results and comparison"
    main->>main: "write_json_results()"
    main->>PlotScript: "Execute: python3 plot_results.py"
    PlotScript->>PlotScript: "Read JSON results"
    PlotScript->>PlotScript: "Generate timing_over_time.png"
    PlotScript->>PlotScript: "Generate simple_histograms.png"
    PlotScript-->>main: "Plots saved"
    main-->>User: "Display results and plot locations"
```

<!-- greptile_other_comments_section -->

<!-- /greptile_comment -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Benchmark converted to a native C++ executable with configurable scheduling, CPU pinning, workloads, installable runtime, and updated invocation.
  * Added a plotting utility to generate timing-over-time and histogram visualizations from benchmark output.

* **Documentation**
  * README refreshed with new configuration options, example output, architecture/timing guidance, troubleshooting, and plot/output instructions.

* **Chores**
  * Removed the prior Python benchmark implementation and updated metadata to reflect the compiled binary.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nvcyc committed:
Update real-time thread benchmark

- Add comprehensive real-time thread benchmark implementation
- Simplify benchmark structure and remove unused files
- Reformat output logs for better readability
- Add argument for specifying dummy load number

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Update result printout

* Focus on Frame Period Std in result print out in both cpp and python
  version
* Updated readme for cpp usage to use ./holohub

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Update benchamark print out

* Add min/max into benchmark results
* Add target frame period in configuration print out for easier comparison
* Remove percentile average calculation in cpp version

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Update args in realtime threads cpp benchmark

* Update load-intensity arg to bg-load-intensity, bg-workload-size
  bm-load-intensity and bm-workload-size (bg as background, bm as benchmark)
* Update duration in cpp version to 30 as --help mentioned default value
* Update sched_deadline in python to be 100% of period to match cpp

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Update dummyload in python version to match cpp version

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Add automatic plot generation to C++ realtime thread benchmark

- Create plot_results.py script to generate timing analysis plots from JSON output
- Integrate automatic plotting into C++ benchmark after JSON generation
- Add CMake configuration to copy plotting script to build directory
- Generate same plots as Python version: timing_over_time.png and simple_histograms.png
- Enable single-command workflow: C++ benchmark now produces both console output and visual plots
- Maintain compatibility with existing JSON output format

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Simplify realtime_threads_benchmarking to C++ only implementation

- Remove Python version and consolidate to single C++ implementation
- Move C++ files from cpp/ subdirectory to main directory
- Remove python/ directory and all Python-specific files
- Update CMakeLists.txt to reflect flattened directory structure
- Maintain single metadata.json for C++ implementation
- Update README.md to remove version-specific language and dual implementation references
- Preserve all functionality: automatic plotting, parameter configuration, and JSON output
- Update images in README

The benchmark now provides a streamlined single-language experience while
maintaining all features including automatic plot generation and comprehensive
real-time scheduling analysis.

Directory structure change:
- Before: realtime_threads_benchmarking/{cpp/,python/}
- After: realtime_threads_benchmarking/ (flattened)

Usage simplified:
- Before: ./holohub run realtime_threads_benchmarking --language=cpp
- After: ./holohub run realtime_threads_benchmarking

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Update graph and result based on RTX 6000 Ada Generation machine

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Fix lint issues

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

tbirdso commented:
Discussed with @sohamm17 , review is in progress and comments will be available in a couple of weeks.

sohamm17 reviewed (changes requested):
Thanks for converting this example to cpp! Left some comments to address.

chengronglai committed:
Refactor: Add getopt_long, CPU pinning, and microsecond precision

- Replace manual argument parsing with standard getopt_long
- Add --pin-cpu/-c option for configurable CPU affinity
- Increase precision to 6 decimals for sub-millisecond measurements
- Refactor JSON writing to eliminate code duplication

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

sohamm17 reviewed (commented):
Looks good except a minor issue. Thanks!

chengronglai committed:
Use sysconf for Linux CPU count detection

Replace hardware_concurrency() with sysconf(_SC_NPROCESSORS_CONF) for
accurate CPU count matching kernel CPU numbering.

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

sohamm17 reviewed (approved):
LGTM

chengronglai committed:
Fix issues for edge cases

* Prevent division-by-zero calculation
* Add exception handling when open JSON data
* Add a check for plot script before generating plots
* Add checks when calculating bins_period based on raw data,
  prevent calling min/max on empty data

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai commented on benchmarks/realtime_threads_benchmarking/plot_results.py:
```diff
@@ -0,0 +1,268 @@
+        min(min(normal_periods), min(realtime_periods)) * 0.9,
+        max(max(normal_periods), max(realtime_periods)) * 1.1,
+        150,  # Finer resolution
+    )
```
done

chengronglai committed:
Fix critical bugs in realtime thread benchmark

* Add missing standard headers
* Add unsorted raw data in JSON for time-series plots
* Ensure dummy workload isn’t optimized away
* Create output directory before writing JSON
* Update histogram and time-series plot

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai commented on benchmarks/realtime_threads_benchmarking/realtime_thread_benchmark.cpp:
```diff
@@ -0,0 +1,659 @@
+  double max_val = 0.0;
+  size_t sample_count = 0;
+  std::vector<double> sorted_data;
+};
```
done

chengronglai commented on benchmarks/realtime_threads_benchmarking/realtime_thread_benchmark.cpp:
```diff
@@ -0,0 +1,659 @@
+#include <unistd.h>
+
+#include <holoscan/holoscan.hpp>
+
```
done

chengronglai committed:
Add input validation and remove unused members

* Validate target-fps and duration parameters
* Remove unused target_period_ns_ member from BenchmarkOp

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

bhashemian committed:
Merge branch 'main' into benchmarks/rt_threads_cpp

chengronglai committed:
Remove unused members and Eliminate command injection

* Replace system() with posix_spawnp (no shell)
* Remove unused _lock unused member

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Merge branch 'main' into benchmarks/rt_threads_cpp

chengronglai commented on benchmarks/realtime_threads_benchmarking/CMakeLists.txt:
```diff
@@ -13,10 +13,31 @@
-)
\ No newline at end of file
+find_package(holoscan 3.5 REQUIRED CONFIG
+             PATHS "/opt/nvidia/holoscan" "/workspace/holoscan-sdk/install")
```
Consistent with the codebase

chengronglai commented on benchmarks/realtime_threads_benchmarking/metadata.json:
```diff
@@ -7,7 +7,7 @@
+        "language": "cpp",
         "version": "1.0",
         "changelog": {
             "1.0": "Initial Release"
```
OK to leave it as 1.0

chengronglai commented on benchmarks/realtime_threads_benchmarking/realtime_thread_benchmark.cpp:
```diff
@@ -0,0 +1,701 @@
+      work_result += std::sqrt(std::abs(x) + 1.0);
+    }
+  }
+  return work_result;
```
verified changing workload-size and load-intensity shows different results, proving the load is working

chengronglai committed:
Fix variable shadowing in run_dummy_cpu_workload

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai commented on benchmarks/realtime_threads_benchmarking/realtime_thread_benchmark.cpp:
```diff
@@ -0,0 +1,701 @@
+  std::vector<double> data(workload_size);
+  double work_result = 0.0;
+  for (int i = 0; i < load_intensity; ++i) {
+    for (size_t i = 0; i < data.size(); ++i) {
```
done

bhashemian committed:
Merge branch 'main' into benchmarks/rt_threads_cpp
